Command Line coloring

morm24 opened the pull request:
# Description

Added an interface to enable coloring the console output.
Important are the classes Colorizer (Colorizer.ccp & Colorizer.h).
Changes in TOPPBasse and other classes are just for testing and not final in any way right now. 


# Checklist:
- [x] Make sure that you are listed in the AUTHORS file
- [-] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [-] New and existing unit tests pass locally with my changes
- [-] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.
- Or click on the action: e.g., for clang-format linting

# Note:
- Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

# Advanced commands (admins / reviewer only):
- `/reformat` (experimental) applies the clang-format style changes as additional commit
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds

Commit:
Added class and header Colorizer.* into concept. Basic functionality for colorizing command line output (on Linux) is esablished

morm24 committed:
Merge branch 'OpenMS:develop' into coloring

Commit:
updated colorizer class

Commit:
Coloring aktualisiert, lokases verzeichnis auf neusten stand gebracht.

Commit:
added full functionality for windows use

Commit:
added colorizer to sources.cmake

Commit:
colorizer in TOPPBase hinzugefügt, Testcodes eingefügt

Commit:
changed to colorizing functions, with std::string output rather than Colorizer Objects. (enables colorizer calls in functions with std::string expection)

Commit:
added few comments, clead functions, ready for Pull request (overview)

hbeschorner commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+    
+    
+    
+    template <typename T = std::string>
```
why a template if it's always a string

morm24 replied:
its not always a string, the default valueis a string. It has to accept all types, that work with the << operator. 
I had to define the default type allready in the Template. Just writing the default in the Attributes wouldn't be accepted.

hbeschorner commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+
+  private:
+    //std::string _text;
+    const int _color;
```
```suggestion
    const int color_;
```

hbeschorner commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+  private:
+    //std::string _text;
+    const int _color;
+    std::stringstream _input;
```
```suggestion
    std::stringstream input_;
```

hbeschorner commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+      //clear was not possible (resets some flags in the sream)
+      //std::string str = "";
+      this->_input.str("");
+      this->_input << s;
```
maybe
```suggestion
      this->_input.str(s);
```

cbielow replied:
```suggestion
      this->_input.str(""); // clear the stringstream
      this->_input << s;  // add new data
```

hbeschorner commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+    //std::string _text;
+    const int _color;
+    std::stringstream _input;
+    bool reset = true;
```
```suggestion
    bool reset_ = true;
```

hbeschorner commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+    std::stringstream _input;
+    bool reset = true;
+#if defined(_WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(_WIN64)
+    int _defcolor;
```
```suggestion
    int default_color_;
```

hbeschorner commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+  };
+
+
+//deklaration of all colorizer object. 
```
```suggestion
//declaration of all colorizer object. 
```

hbeschorner commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+  {
+    std::stringstream text;
+    text << color_black(s);
+      return text.str(); 
```
```suggestion
      return color_black(s); 
```
same with the other ones

morm24 replied:
doesn't word, bc. there is the return expects a std::string, the Object (color_black(s)) is an object of type colorizer, with no automatic cast to std::string. 
thats why I use the workaround, with the overloaded operator <<, so i can return a std::string object, casted from the  std::stringstream object.

hbeschorner commented on src/openms/source/APPLICATIONS/TOPPBase.cpp:
```diff
@@ -1368,7 +1370,7 @@ namespace OpenMS
                     "Either provide a full filepath or fix your PATH environment!" +
                     (p.required ? "" : " Since this file is not strictly required, you might also pass the empty string \"\" as "
-                    "argument to prevent its usage (this might limit the usability of the tool)."));
+                    "argument to prevent it's usage (this might limit the usability of the tool)."));
```
pretty sure its was correct

morm24 replied:
ddint know I changed it. yes, old one is correct

lennykovac commented on src/openms/source/CONCEPT/Colorizer.cpp:
```diff
@@ -0,0 +1,173 @@
+    //If we are on windows, get the "default" console color and safe it in _defcolor
+    #if defined(_WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(_WIN64)
+    CONSOLE_SCREEN_BUFFER_INFO Info;
+    HANDLE hStdout = GetStdHandle(STD_OUTPUT_HANDLE);
```
weird var-name
```suggestion
    HANDLE std_out_handler = GetStdHandle(STD_OUTPUT_HANDLE);
```

lennykovac commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+    
+    auto getColor(int i = -1);
+    std::string getText();
+    bool getReset() {return this->reset;}
```
Coding-Convention?
```suggestion
    bool getReset() 
    {
      return this->reset;
    }
```

hbeschorner commented on src/openms/source/CONCEPT/Colorizer.cpp:
```diff
@@ -0,0 +1,173 @@
+  /// Default destructor
+  Colorizer::~Colorizer()
+  {
+    //if colorizer oibject is destroyed, set console color back to def col. 
```
```suggestion
    //if colorizer object is destroyed, set console color back to def col. 
```

lennykovac commented on src/openms/source/CONCEPT/Colorizer.cpp:
```diff
@@ -0,0 +1,173 @@
+#if defined(_WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(_WIN64)
+   
+    //set color of output   
+    SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE),col.getColor());
```
```suggestion
    SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), col.getColor());
```

lennykovac reviewed (commented):
Looks good overall but stick to the coding-convention and a lot of unnecessary spaces/newlines.
Try to stick to english-comments :)

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Moritz Berger ; Tetana Krymovska$
+// $Authors: Moritz Berger ; Tetana Krymovska$
```
```suggestion
// $Maintainer: Moritz Berger, Tetana Krymovska$
// $Authors: Moritz Berger, Tetana Krymovska$
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+{
+
+  /**
+   * @brief A class, that provides options for coloring String output with the "<<" Operator
```
```suggestion
   * @brief A class, that provides options for colored output with the "<<" operator for output streams (cout, cerr)
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+    /** @name Constructors and Destructor
+     */
+    //@{
+    Colorizer(int color); 
```
```suggestion

  enum class COLOR  // todo: move this outside of class
  {
     black,
     red,
     green,
     yellow,
     blue,
     magenta,
     cyan,
     white,
     RESET, ///< reset the color to the previous (default?) color
  };

    /** @name Constructors and Destructor
     */
    //@{
    Colorizer(COLOR color); 
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+    /// Copy constructor
+    //Colorizer(const Colorizer &rhs);
+
+    /// Default destructor
```
```suggestion
    /// Destructor
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+    Colorizer(int color); 
+
+    /// Copy constructor
+    //Colorizer(const Colorizer &rhs);
```
make explicit
= delete
or
= default

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+      this->_input << s;
+     
+
+      (this->_input.str() == "")? this->reset = false: this->reset = true;
```
```suggestion
     reset_ = true;
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+    
+    
+    template <typename T = std::string>
+    Colorizer& operator()(T s = T(""))
```
```suggestion

    Colorizer& operator()()
    {
      reset_ = false;
      this->_input.str(""); // clear the stream 
      return *this;
    }
    
    
    template <typename T>
    Colorizer& operator()(T s)
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+    const int _color;
+    std::stringstream _input;
+    bool reset = true;
+#if defined(_WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(_WIN64)
```
```suggestion
#ifdef OPENMS_WINDOWSPLATFORM
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+     *
+     */
+#if defined(__linux__) || defined(__OSX__)
+    inline static const std::array<const char*,9> colors {"\033[30m", "\033[31m", "\033[32m", "\033[33m", "\033[34m", "\033[35m", "\033[36m", "\033[37m", "\033[0m"};
```
```suggestion
    inline static constexpr std::array<const char*,9> colors_ {"\033[30m", "\033[31m", "\033[32m", "\033[33m", "\033[34m", "\033[35m", "\033[36m", "\033[37m", "\033[0m"};
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+#if defined(__linux__) || defined(__OSX__)
+    inline static const std::array<const char*,9> colors {"\033[30m", "\033[31m", "\033[32m", "\033[33m", "\033[34m", "\033[35m", "\033[36m", "\033[37m", "\033[0m"};
+#elif defined(_WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(_WIN64)
+    inline static const std::array<const int,9> colors {16,12,10,14,9,13,11,15,15};
```
```suggestion
    inline static constexpr std::array<const int,9> colors_ {16,12,10,14,9,13,11,15,15};
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+
+
+//deklaration of all colorizer object. 
+  extern /*OPENMS_DLLAPI*/ Colorizer color_black;
```
```suggestion
  extern /*OPENMS_DLLAPI*/ Colorizer make_black;
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,248 @@
+  extern /*OPENMS_DLLAPI*/ Colorizer make_magenta;
+  extern /*OPENMS_DLLAPI*/ Colorizer make_cyan;
+  extern /*OPENMS_DLLAPI*/ Colorizer make_white;
+  extern /*OPENMS_DLLAPI*/ Colorizer make_def;
```
```suggestion
  extern /*OPENMS_DLLAPI*/ Colorizer reset_color; ///< reset the color to default, alias for 'make_default_color'
  extern /*OPENMS_DLLAPI*/ Colorizer make_default_color; ///< reset the color to default, alias for 'reset_color'
  
```

cbielow commented on src/openms/source/APPLICATIONS/TOPPBase.cpp:
```diff
@@ -40,6 +40,7 @@
 
 #include <OpenMS/CONCEPT/LogStream.h>
 #include <OpenMS/CONCEPT/VersionInfo.h>
+#include <OpenMS/CONCEPT/Colorizer.h>
```
order includes

cbielow commented on src/openms/source/CONCEPT/Colorizer.cpp:
```diff
@@ -0,0 +1,173 @@
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Moritz Berger; Tetana Krymovska $
+// $Authors: Moritz Berger ; Tetana Krymovska$
```
```suggestion
// $Maintainer: Moritz Berger, Tetana Krymovska $
// $Authors: Moritz Berger, Tetana Krymovska$
```

cbielow commented on src/openms/source/CONCEPT/Colorizer.cpp:
```diff
@@ -0,0 +1,173 @@
+// $Authors: Moritz Berger ; Tetana Krymovska$
+// --------------------------------------------------------------------------
+
+#include </buffer/ag_bsc/pmsb_22/morib70/openms/OpenMS/src/openms/include/OpenMS/CONCEPT/Colorizer.h>
```
```suggestion
#include <OpenMS/CONCEPT/Colorizer.h>
```

cbielow commented on src/openms/source/CONCEPT/Colorizer.cpp:
```diff
@@ -0,0 +1,173 @@
+  // constructor
+  Colorizer::Colorizer(int color): _color(color)
+  {
+    //If we are on windows, get the "default" console color and safe it in _defcolor
+    #if defined(_WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(_WIN64)
+    CONSOLE_SCREEN_BUFFER_INFO Info;
+    HANDLE hStdout = GetStdHandle(STD_OUTPUT_HANDLE);
+    GetConsoleScreenBufferInfo(hStdout, &Info);
+    _defcolor = Info.wAttributes;
```
remember two default colors, for cerr and cout

cbielow replied:
move this to another class (construction + destrcution)
```

namespace OpenMS
{

namespace  Internal
{
struct WindowsOSDefaultColor
{
  WindowsOSDefaultColor()
 {
   std::cout << "saving default colors...\n";
   /// get and remember 2 default colors
}
 ~WindowsOSDefaultColor()
{
  std::cout << "restoring default colors...\n";
}

int default_cerr_;
int default_cout_;
}; // end WindowsOSDefaultColor

extern static const WindowsOSDefaultColor default_color____;

} // internal
} // OpenMS

cbielow commented on src/openms/source/CONCEPT/Colorizer.cpp:
```diff
@@ -0,0 +1,173 @@
+   * 8=default console color (reset)
+  \return std::string contaaining ANSI escape characters for colors
+  */
+  auto Colorizer::getColor(int i)
+  {
+    //checking if parameter is outside array size or not set at all
+    // Darf nicht NULL, da dadurch kein "black" mehr möglich ist
+    return ((i == -1 || i > 8 || i < 0) ? this->colors[this->_color] : this->colors[i]);
+  }
+
+  std::string Colorizer::getText()
+  { 
+
+    return _input.str();
+  }   
```
delete

cbielow commented on src/openms/source/CONCEPT/Colorizer.cpp:
```diff
@@ -0,0 +1,173 @@
+
+    return _input.str();
+  }   
+  // overload the shift operator (<<)
+  std::ostream &operator<<(std::ostream &o_stream, OpenMS::Colorizer& col)
```
```suggestion
  // overload the shift operator (<<)
  std::ostream &operator<<(std::ostream &o_stream, OpenMS::Colorizer& col)
  {
    col.outputToStream(o_stream);
    return o_stream;
  }
```

cbielow commented on src/openms/source/CONCEPT/Colorizer.cpp:
```diff
@@ -0,0 +1,173 @@
+  // overload the shift operator (<<)
+  std::ostream &operator<<(std::ostream &o_stream, OpenMS::Colorizer& col)
+  {
+    // colorize string with color set in the object
+#if defined(_WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(_WIN64)
+   
+    //set color of output   
+    SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE),col.getColor());
+
+    // paste text
+    o_stream << col.getText();
+    
+    // recover old Console font and set it as new one.
+    if(col.getReset()) 
+    {
+    SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), col._defcolor);
```
move to `Colorizer::outputToStream(std::ostream&);`

morm24 commented on src/openms/source/APPLICATIONS/TOPPBase.cpp:
```diff
@@ -1368,7 +1370,7 @@ namespace OpenMS
                     "Either provide a full filepath or fix your PATH environment!" +
                     (p.required ? "" : " Since this file is not strictly required, you might also pass the empty string \"\" as "
-                    "argument to prevent its usage (this might limit the usability of the tool)."));
+                    "argument to prevent it's usage (this might limit the usability of the tool)."));
```
```suggestion
                    "argument to prevent its usage (this might limit the usability of the tool)."));
```

morm24 committed:
Apply suggestions from code review

Changes made in pull request code review. Not working, will be fixed by next commit.

Co-authored-by: Chris Bielow <chris.bielow@fu-berlin.de>
Co-authored-by: Lenny Kovac <44415793+lennykovac@users.noreply.github.com>
Co-authored-by: hbeschorner <102038306+hbeschorner@users.noreply.github.com>

Commit:
changed pull request changes, now working with linux. windows functionality will be set with next commit

Commit:
corrected include path, addition to last commit: removed ability to color inside function call

morm24 committed:
Error in windows functionality: struct WindowsOSDefaultColor when compiling.

Commit:
struct error should be fixed now.

Commit:
updatetd colorizer.cpp include path

morm24 committed:
static WindowsOSDefaultColor struct now working under Windows (still verbose).

Commit:
work in progress, started changing the Console Utils, tidying up Colorizer.

Commit:
Colorizer functionality now working again and tested with linux (windows not tested). Basics of IndentedStringStream implemented.

Commit:
kompiliert, fehler in FOrmatierung, und zugriffsfehler innerhalb der breakstring func. keine Testklassen

cbielow commented on src/openms/include/OpenMS/APPLICATIONS/ConsoleUtils.h:
```diff
@@ -35,46 +35,158 @@
+      
+      OpenMS::StringList result = ConsoleUtils::breakString(string_to_print,indentation_,max_lines_,current_column_pos_);
+      
+      current_column_pos_ = result.back().size() -1;
```
```suggestion
      if (result.size()>=2)
      { // we completed the previous line, so start counting from the latest incomplete line
        current_column_pos_ = result.back().size();
       }
       else
       { // only one line; simply forward the column position
         current_column_pos_ += result.back().size();
       }
```

cbielow commented on src/openms/include/OpenMS/APPLICATIONS/ConsoleUtils.h:
```diff
@@ -35,46 +35,158 @@
+      
+
+
+        *stream_ << ListUtils::concatenate(result," ");
```
```suggestion
        *stream_ << ListUtils::concatenate(result, '\n');
```

cbielow commented on src/openms/include/OpenMS/CONCEPT/Colorizer.h:
```diff
@@ -0,0 +1,163 @@
+    Colorizer& operator()()
+    {
+      reset_ = false;
+      this->input_.str(""); // clear the stream
```
```suggestion
      this->input_.str(""); // clear the stringstream
```

cbielow commented on src/openms/source/APPLICATIONS/ConsoleUtils.cpp:
```diff
@@ -54,16 +54,54 @@ namespace OpenMS
+  ConsoleUtils::~ConsoleUtils()
+  {
+#ifdef OPENMS_WINDOWSPLATTFORM
+    setCoutColor(default_cout_);
+    setCerrColor(default_cerr_);
```
```suggestion
    resetConsoleColor();
```

cbielow commented on src/openms/source/APPLICATIONS/ConsoleUtils.cpp:
```diff
@@ -54,16 +54,54 @@ namespace OpenMS
+#endif 
   }
 
   ConsoleUtils::ConsoleUtils(ConsoleUtils const& other) :
     console_width_(other.console_width_)
   {
+#ifdef OPENMS_WINDOWSPLATTFORM
+    default_cout_ = other.default_cout_;
+    default_cerr_ = other.default_cerr_;
+#endif
   }
 
   void ConsoleUtils::operator=(const ConsoleUtils& other)
   {
     console_width_ = other.console_width_;
+#ifdef OPENMS_WINDOWSPLATTFORM
+    default_cout_ = other.default_cout_;
+    default_cerr_ = other.default.cerr_;
+#endif 
```
```suggestion
  ConsoleUtils::ConsoleUtils(ConsoleUtils const& other) = default;
  void ConsoleUtils::operator=(const ConsoleUtils& other) = default;
```

cbielow commented on src/openms/source/APPLICATIONS/ConsoleUtils.cpp:
```diff
@@ -206,7 +258,48 @@ namespace OpenMS
+void setCoutColor(int color_code)
+{
+  SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), color_code);
+  
```
```suggestion
```

morm24 committed:
Update src/openms/include/OpenMS/APPLICATIONS/ConsoleUtils.h

Co-authored-by: Chris Bielow <chris.bielow@fu-berlin.de>

morm24 committed:
Update src/openms/include/OpenMS/APPLICATIONS/ConsoleUtils.h

Co-authored-by: Chris Bielow <chris.bielow@fu-berlin.de>

morm24 committed:
Update src/openms/include/OpenMS/CONCEPT/Colorizer.h

Co-authored-by: Chris Bielow <chris.bielow@fu-berlin.de>

morm24 committed:
Update src/openms/source/APPLICATIONS/ConsoleUtils.cpp

Co-authored-by: Chris Bielow <chris.bielow@fu-berlin.de>

morm24 committed:
Update src/openms/source/APPLICATIONS/ConsoleUtils.cpp

Co-authored-by: Chris Bielow <chris.bielow@fu-berlin.de>

morm24 committed:
Update src/openms/source/APPLICATIONS/ConsoleUtils.cpp

Co-authored-by: Chris Bielow <chris.bielow@fu-berlin.de>
